ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/StateMachine.hs

```diff
@@ -295,44 +298,71 @@ type TestIterator m blk = WithEq (Iterator m blk (AllComponents blk))
 -- | Short-hand
 type TestReader   m blk = WithEq (Reader   m blk (AllComponents blk))
 
+-- | The current ChainDB instance and things related to it.
+--
+-- When closing and reopening the ChainDB, this record will be replaced in the
+-- 'varDB' field of 'ChainDBEnv' with a new one.
+data ChainDBState m blk = ChainDBState
+    { chainDB       :: ChainDB m blk
+    , internal      :: ChainDB.Internal m blk
+    , addBlockAsync :: Async m Void
+      -- ^ Background thread that adds blocks to the ChainDB
+    }
+  deriving NoUnexpectedThunks via AllowThunk (ChainDBState m blk)
+
 -- | Environment to run commands against the real ChainDB implementation.
 data ChainDBEnv m blk = ChainDBEnv
-  { chainDB    :: ChainDB m blk
-  , internal   :: ChainDB.Internal m blk
+  { varDB      :: StrictMVar m (ChainDBState m blk)
   , registry   :: ResourceRegistry m
   , varCurSlot :: StrictTVar m SlotNo
   , varNextId  :: StrictTVar m Id
   , varVolDbFs :: StrictTVar m MockFS
+  , args       :: ChainDbArgs m blk
+    -- ^ Needed to reopen a ChainDB, i.e., open a new one.
   }
 
-run :: forall m blk. (IOLike m, HasHeader blk)
+-- PRECONDITION: the ChainDB is closed
+reopen :: (IOLike m, LedgerSupportsProtocol blk) => ChainDBEnv m blk -> m ()
```

Comment:
We decide to not go for this approach, as the current tests support running a command against a closed database, which is impossible with the proposed approach. It would also require special casing the `(Re)Open` command.